### ISSUE TITLE
FLINT uses reserved identifiers extensively

### ISSUE THREAD

**Issue by jscott0** [wbhart/flint2#1136]
[A good source on the topic is here](https://wiki.sei.cmu.edu/confluence/display/c/DCL37-C.+Do+not+declare+or+define+a+reserved+identifier), but FLINT uses many identifiers reserved by ISO C and POSIX. Defining such an identifier ourselves causes undefined behavior (this is how feature test macros such as `_GNU_SOURCE` and `_POSIX_C_SOURCE` are able to work). Here are some examples:
- functions starting with an underscore
- identifiers starting with two underscores (`__FLINT_VERSION`, `__flint_set_memory_functions`)
- types ending in `_t` (`flint_rand_t`, POSIX only)

As well as infecting any FLINT-using application with undefined behavior (since applications generally need to use the same identifiers as FLINT does, e.g. functions and type names), this can also restrict portability, as any such identifier may be pre-defined and hence lead to a name collision.

Obviously this issue cannot be solved overnight, but I'd like to bring it to your attention so interfaces can be namespace-clean going forward.

**Comment by wbhart** [comment]
I sincerely doubt anyone is going to pay any attention to a standard that disallows underscores or type names ending in _t. This is commonplace in libraries that have existed for decades longer than the C standard in question.

In addition, I tried to access the cited reference for this, but it does not provide me with a link, so I am not even able to verify if this is in said standard, which is hardly satisfactory in 2022.

There are projects of millions of lines of code that rely on libraries which pervasively name types with _t (e.g. GMP, MPFR, MPFC, MPIR, the list goes on ad infinitum).

**Comment by fredrik-johansson** [comment]
Yes, the issue here is just the potential for name conflicts with system libraries, and such conflicts typically result in compiler warnings or errors. I seriously doubt any compiler is going to see an identifier with a leading `__` or a trailing `_t` and silently emit wrong code out of spite. In any case, we already run into occasional name conflicts with names that *don't* violate these guidelines (see workarounds for ``ulong`` for example), which is the price to pay in a language without proper namespaces. The solution is just to avoid common names as far as possible. I see little reason to worry about ``__FLINT_VERSION``, but we might want to use the ``flint_`` and ``FLINT_`` prefixes in more places.

**Comment by albinahlback** [comment]
I vote for closing this as I don't see that this ever being resolved.

**Comment by wbhart** [comment]
It's an issue if Flint ever uses C11. Eventually compilers will switch to that as the default standard,

But I agree. As far as Flint is concerned, there is not the motivation in the community to do anything about this for the foreseeable future. And so long as C compilers offer a -ansi or even -std=c99 option, I don't think this will affect Flint.
